src/main/java/net/wasdev/wlp/common/springboot/util/SpringBootThinUtil.java

```diff
@@ -20,183 +25,433 @@
-				writeEntry(is, thinJar, path);
-			}
-		}
-	}
+    private void thin() throws FileNotFoundException, IOException, NoSuchAlgorithmException {
+        try (JarOutputStream thinJar = new JarOutputStream(new FileOutputStream(targetThinJar), sourceFatJar.getManifest())) {
+            Set<String> entryNames = new HashSet<>();
+            for (Enumeration<JarEntry> entries = sourceFatJar.entries(); entries.hasMoreElements();) {
+                JarEntry entry = entries.nextElement();
+                if (entryNames.add(entry.getName()) && !JarFile.MANIFEST_NAME.equals(entry.getName()) &&
+                    !entry.getName().startsWith(SPRING_BOOT_LOADER_CLASSPATH) /* omit spring boot loader classes */) {
+                    storeEntry(thinJar, entry);
+                }
+            }
+            addLibIndexFileToThinJar(thinJar);
 
-	private static boolean isEmbeddedContainerImpl(JarEntry entry) {
-		return isEmbeddedContainerImpl(entry.getName());
-	}
+        }
+    }
 
-	public static boolean isEmbeddedContainerImpl(String jarName) {
-		String lowerCaseName = jarName.toLowerCase();
-		return (lowerCaseName.endsWith(".jar") && lowerCaseName.contains("tomcat-"));
-	}
+    private void storeEntry(JarOutputStream thinJar, JarEntry entry) throws IOException, NoSuchAlgorithmException {
+        String path = entry.getName();
+        boolean isLibPath = isFromLibPath(path);
+        boolean isLibProvidedPath = isFromLibProvidedPath(path);
+        // check if entry is dependency jar or application class
+        if (isLibPath || isLibProvidedPath) {
+            if (!starterFilter.apply(entry.getName()) && (!isLibProvidedPath || includeLibProvidedPaths())) {
+                String hash = hash(sourceFatJar, entry);
+                String hashPrefix = hash.substring(0, 2);
+                String hashSuffix = hash.substring(2, hash.length());
 
-	protected String hash(JarFile jf, ZipEntry entry) throws IOException, NoSuchAlgorithmException {
-		InputStream eis = jf.getInputStream(entry);
-		MessageDigest digest = MessageDigest.getInstance("sha-256");
-		byte[] buffer = new byte[4096];
-		int read = -1;
-
-		while ((read = eis.read(buffer)) != -1) {
-			digest.update(buffer, 0, read);
-		}
-		byte[] digested = digest.digest();
-		return convertToHexString(digested);
-	}
+                storeLibraryInDir(entry, hashPrefix, hashSuffix);
 
-	private static String convertToHexString(byte[] digested) {
-		StringBuilder stringBuffer = new StringBuilder();
-		for (int i = 0; i < digested.length; i++) {
-			stringBuffer.append(Integer.toString((digested[i] & 0xff) + 0x100, 16).substring(1));
-		}
-		return stringBuffer.toString();
-	}
+                String libLine = "/" + path + '=' + hash;
+                libEntries.add(libLine);
+            }
+        } else {
+            try (InputStream is = sourceFatJar.getInputStream(entry)) {
+                writeEntry(is, thinJar, path);
+            }
```

Comment:
Does the InputStream 'is' need to be closed in a finally block?

Reply:
Not required, since this is "try with resource" construct. When the try block finishes the InputStream will be closed automatically.